net-tools-1.45/man/en_US.88591/route.8
.TH ROUTE 8 "8 August 1997" "net-tools" "Linux Programmer's Manual"
.SH NAME
route \- show / manipulate the IP routing table
.SH SYNOPSIS
.B route
.RB [ \-CFvnee ]
.TP
.B route 
.RB [ \-v ]
.RB [ \-A
family] 
.B add 
.RB [ \-net | \-host ] 
target 
.RB [ netmask 
Nm] 
.RB [ gw 
Gw] 
.RB [ metric 
N] 
.RB [ mss 
M] 
.RB [ window 
W] 
.RB [ irtt 
I]
.RB [ reject ]
.RB [ mod ]
.RB [ dyn ] 
.RB [ reinstate ] 
.RB [[ dev ] 
If]
.TP
.B route 
.RB [ \-v ] 
.RB [ \-A
family]
.B del 
.RB [ \-net | \-host ] 
target 
.RB [ gw 
Gw] 
.RB [ netmask 
Nm] 
.RB [ metric 
N] 
.RB [[ dev ]
If]
.TP
.B route 
.RB [ \-V ] 
.RB [ \-\-version ]
.RB [ \-h ]
.RB [ \--help ]
.SH DESCRIPTION
.B Route
manipulates the kernel's IP routing table.  Its primary use is to set
up static routes to specific hosts or networks via an interface after
it has been configured with the
.BR ifconfig (8)
program.


.SH OPTIONS
.TP
.B \-v
select verbose operation.

.TP
.B \-A family
Use the specified address family (eg `inet', `inet6').

.TP
.B \-n
show numerical addresses instead of trying to determine symbolic host
names. This is useful if you are trying to determine why the route to your
nameserver has vanished.

.TP
.B \-e
use
.BR netstat (8)-format
for displaying the routing table.
.B \-ee 
will generate a very long line with all parameters from the routing table.

.TP
.B \-net
the
.B target
is a network.

.TP
.B -host
the
.B target 
is a host.

.TP 
.B -F
displays the kernel FIB routing table. The layout can be changed with
.B \-e
and
.B \-ee
.

.TP 
.B -C
displays the kernel's route cache. 

.TP
.B del
deletes a route.

.TP 
.B add 
adds a route.

.TP
.B target
The destination network or host. You can provide IP addresses in dotted
decimal or host/network names.

.TP
.B netmask Nm
modifier specifies the netmask of the route to be added.

.TP
.B gw Gw
Any IP packets for the target network/host will be routed through the 
specified gateway. 
.B NOTE:
The specified gateway must be reachable first. This usually means that
you have to set up a static route to the gateway beforehand. If you specify
the adress of one of your local interfaces, it will be used to decide about
the interface to which the packets should be routed to. This is a BSDism
compatibility hack.

.TP
.B metric M
Set the metric field in the routing table (used by routing daemons) to M.

.TP 
.B mss M
Set the TCP Maximum Segment Size (MSS) for connections over this route
to M bytes. This is normally used only for fine optimisation of
routing setups. The default is 536.

.TP 
.B window W 
Set the TCP window size for connections over this route to W
bytes. This is typically only used on AX.25 networks and with drivers
unable to handle back to back frames.

.TP
.B irtt I
Set the initial round trip time (irtt) for TCP connections over this
route to I milliseconds (1-12000). This is typically only used on
AX.25 networks. If omitted the RFC 1122 default of 300ms is used.

.TP
.B reject
Install a blocking route, which will force a route lookup to fail.
This is for example used to mask out networks before using the default
route.  This is NOT for firewalling.

.TP
.B mod, dyn, reinstate
Install a dynamic or modified route. Both flags are generally only
set by a routing daemon. This is only for diagnostic purpose.

.TP
.B dev If
Forces the route to be associated with the specified device, as the
kernel will otherwise try to determine the device on its own (by
checking already existing routes and device specifications, and where
the route is added to). In most normal networks you won't need this.

If 
.B dev If
is the last option on the command line, the word 
.B dev
may be omitted, as it's the default. Otherwise the order of the route
modifiers (metric - netmask - gw - dev) doesn't matter.

.SH EXAMPLES
.TP
.B route add -net 127.0.0.0
adds the normal loopback entry, using netmask 255.0.0.0 (class A net,
determined from the destination address) and associated with the 
"lo" device (assuming this device was prviously set up correctly with
.BR ifconfig (8)). 

.TP 
.B route add -net 192.56.76.0 netmask 255.255.255.0 dev eth0
adds a route to the network 192.56.76.x via 
"eth0". The Class C netmask modifier is not really necessary here because
192.* is a Class C IP address. The word "dev" can be omitted here. 

.TP
.B route add default gw mango-gw
adds a default route (which will be used if no other route matches).
All packets using this route will be gatewayed through "mango-gw". The
device which will actually be used for that route depends on how we
can reach "mango-gw" - the static route to "mango-gw" will have to be
set up before. 

.TP
.B route add ipx4 sl0
Adds the route to the "ipx4" host via the SLIP interface (assuming that
"ipx4" is the SLIP host).

.TP
.B route add -net 192.57.66.0 netmask 255.255.255.0 gw ipx4
This command adds the net "192.57.66.x" to be gatewayed through the former
route to the SLIP interface.

.TP
.B route add 224.0.0.0 netmask 240.0.0.0 dev eth0
This is an obscure one documented so people know how to do it. This sets
all of the class D (multicast) IP routes to go via "eth0". This is the
correct normal configuration line with a multicasting kernel. 

.TP
.B route add -net 10.0.0.0 netmask 255.0.0.0 reject
This installs a rejecting route for the private network "10.x.x.x."

.LP
.SH OUTPUT
The output of the kernel routing table is organized in the following columns
.TP
.B Destination     
The destination network or destination host.
.TP
.B Gateway
The gateway address or '*' if none set.
.TP
.B Genmask         
The netmask for the destination net; '255.255.255.255' for a host destination
and '0.0.0.0' for the 
.B default
route.
.TP
.B Flags 
Possible flags are
.br
.B U
(route is
.BR up )
.br
.B H
(target is a
.BR host )
.br
.B G
(use
.BR gateway )
.br
.B R
.RB ( reinstate
route for dynamic routing)
.br
.B D
.RB ( dynamically
installed by daemon or redirect)
.br
.B M
.RB ( modified
from routing daemon or rederict)
.br
.B !
.RB ( reject
route)
.TP
.B Metric 
The 'distance' to the target (usually counted in hops). It is not used by
recent kernels, but may be needed by routing daemons.
.TP
.B Ref    
Number of references to this route. (Not used in the Linux kernel.)
.TP
.B Use
Count of lookups for the route.  Depending on the use of -F and -C this will
be either route cache misses (-F) or hits (-C).
.TP
.B Iface
Interface to which packets for this route will be sent.
.TP
.B MSS 
Default maximum segement size for TCP connections over this route.
.TP
.B Window  
Default window size for TCP connections over this route.
.TP
.B irtt
Initial RTT (Round Trip Time). The kernel uses this to guess about the best
TCP protocol parameters without waiting on (possible slow) answers.
.TP
.B HH (cached only)
The number of ARP entries and cached routes that refer to the hardware
header cache for the cached route. This will be \-1 if a hardware
address is not needed for the interface of the cached route (e.g. lo).
.TP
.B Arp (cached only)
Whether or not the hardware address for the cached route is up to date.
.LP
.SH FILES
.I /proc/net/ipv6_route
.br
.I /proc/net/route
.br
.I /proc/net/rt_cache
.LP
.SH SEE ALSO
.I ifconfig(8), netstat(8), arp(8), rarp(8)
.LP
.SH HISTORY
.B Route
for Linux was originally written by Fred N.  van Kempen,
<waltje@uwalt.nl.mugnet.org> and then modified by Johannes Stille and
Linus Torvalds for pl15. Alan Cox added the mss and window options for
Linux 1.1.22. irtt support and merged with netstat from Bernd Eckenfels.
.SH AUTHOR
Currently maintained by Phil Blundell <Philip.Blundell@pobox.com>.


debian/changelog
update-inetd (4.46) UNRELEASED; urgency=medium

  * Switch git repository to permanent URL, containing all known history
    imported.
  * Add new CHANGES POD section in DebianNet module.
  * Move xinetd handling into DebianNet module.
  * Code cleanup:
    - Declare all variables with our or my.
    - Use a typeglob instead of redefining fallback functions.
    - Modernize print calls, by removing parens, surrounding file handle
      with {}, and using low-precedence operators for die calls.
    - Do not import Debconf::Client::ConfModule module globally, which means
      we will stop getting annoying warnings when testing as non-root.
    - Switch from require to use.
    - Enable strict and warnings pragmas.
    - Get rid of an indentation level by folding an if into an elsif.
    - Do not interpolate string literals.
    - Use Carp instead of warn without newline for an internal error.
    - Convert a print STDERR into a printv.
    - Switch from print STDERR to warn.
    - Use braces around the filehande with print.
    - Make Heredoc terminator quoting style explicit.
    - Use a filehandle variables instead of a barewords.
    - Do not use the two-argument form of open.
    - Uppercase all package variables.
    - Check open() return value.
    - Use braces for regex delimiters.
    - Do not use unless with negative expressions.
    - Do not mix boolean operators of different precedence.
    - Remove useless topic variable usage.
    - Remove regex captures that are unused.
    - Use old decimal instead of dotted-decimal version.
    - Force a scalar context to fix perlcritic false positive.
    - Use List::Util instead of grep in boolean context.
    - Do not use & sigil for function calls.
    - Unpack arguments in printv().
    - Use m{} instead of m// to avoid having to escape /.
    - Stop requiring a newline for printv() calls.
    - Use proper balanced single quotes instead of unbalanced `' pairs.
    - Use low precedence or operators for die fallback call.
    - Use //= instead of "unless (defined $var) { $var = value }".
    - Switch statements to post-conditions.
    - Remove parenthesis for builtins.
    - Surround assignment operator with spaces.
    - Remove trailing semicolon after if block.
    - Bump minimum perl version to 5.010.
    - Prefix private functions with an underscore.
    - Add a space after «my» keywords.

 -- Guillem Jover <guillem@debian.org>  Wed, 01 Aug 2018 10:16:58 +0200

update-inetd (4.45) unstable; urgency=medium

  * Adopt package. (Closes: #719794)
  * Update Vcs fields for move to git.hadrons.org.
  * Line wrap and sort dependency fields.
  * Remove versioned Build-Depends on Essential:yes coreutils, satisfied
    in oldstable.
  * Remove versioned Replaces/Breaks on netbase, satisfied in oldstable.
  * Fix debian/rules target relationship and dependencies.
  * Fix perl dependencies:
    - Remove bogus dependency on libfile-copy-recursive-perl, unused.
    - Remove dependency on libfile-temp-perl, a virtual provided by
      perl-base, which is Essential:yes.
    - Add a dependency on ${perl:Depends} and call dh_perl.
  * Namespace debhelper files with package name.
  * Move debhelper command arguments into debhelper files.
  * Bump debhelper to compatibility level 11.
  * Unify license and copyright file headers.
  * Switch debian/copyright to machine readable format.
  * Bump Standards-Version to 4.1.4 (no changes needed).
  * Rewrite the test suite from python to perl.
  * Convert and merge the man page into inline POD in the perl module.
  * Convert update-inetd man page into POD.
  * Set Rules-Requires-Root to no.

 -- Guillem Jover <guillem@debian.org>  Mon, 14 May 2018 03:16:50 +0200

update-inetd (4.44) unstable; urgency=low

  * Orphan package.
  * Drop README, which points to long-deprecated DEP9 (Closes: #849457).
  * Bump up debhelper compatibility level to 9, and update versioned Build-Dep
    for debhelper.
  * Fix dh-clean-k-is-deprecated and vcs-field-not-canonical lintian warnings.
  * Bump up standards version to 3.9.8 (no changes).

 -- Serafeim Zanikolas <sez@debian.org>  Sun, 15 Jan 2017 11:37:03 +0000

update-inetd (4.43) unstable; urgency=low

  * Add --pattern support for remove mode
  * Update manpage
    - Document support for --pattern and --multi in remove mode
    - Fix argument type for --remove (was documented as ENTRY, whereas it is
      SERVICE)
    - Add copyright notice for myself in manpage

 -- Serafeim Zanikolas <sez@debian.org>  Sun, 03 Jun 2012 00:10:20 +0200

update-inetd (4.42) unstable; urgency=low

  * Declare as Multi-Arch: foreign; thanks to Goswin von Brederlow
    <goswin-v-b@web.de> (Closes: #673398).
  * Add Polish translation; thanks to Michał Kułach" <michalkulach@gmail.com>
    (Closes: #670646).
  * Bump Standards-Version to 3.9.3 (no changes)
  * Add copyright notices for myself in debian/copyright, update-inetd, and
    DebianNet.pm
  * Drop TODO file and add README file, which declares update-inetd as
    deprecated and points to reconf-inetd and DEP9

 -- Serafeim Zanikolas <sez@debian.org>  Sun, 20 May 2012 16:24:48 +0200

update-inetd (4.41) unstable; urgency=low

  * Fall back to external 'tempfile' and 'mv' commands in case perl-base and
    perl-modules are out of sync during an upgrade (Closes: #649174;
    LP: #862129). Thanks to Colin Watson <cjwatson@ubuntu.com> for the bug
    report and patch.

 -- Serafeim Zanikolas <sez@debian.org>  Tue, 22 Nov 2011 23:15:02 +0100

update-inetd (4.40) unstable; urgency=low

  * Fix breakage with non-default inetd packages (Closes: #638180)
  * Add Slovak translation (thanks, Slavko <linux@slavino.sk>;
    Closes: #639449)

 -- Serafeim Zanikolas <sez@debian.org>  Sun, 11 Sep 2011 19:03:14 +0200

update-inetd (4.39) unstable; urgency=low

  * Add Korean translation (thanks, Min-Ji Kang <justminji@gmail.com>;
    closes: #632020).
  * Thanks to Christian Perrier <bubulle@debian.org> for the NMU.

net-tools-1.32-alpha/route.c
/*
 * route        This file contains an implementation of the command
 *              that manages the IP routing table in the kernel.
 *
 * Version:     route 1.91 (1996-04-13)
 *
 * Maintainer:	Bernd 'eckes' Eckenfels, <net-tools@lina.inka.de>
 *
 * Author:      Fred N. van Kempen, <waltje@uwalt.nl.mugnet.org>
 *		(derived from FvK's 'route.c     1.70    01/04/94')
 *
 * Modifications:
 *	 	Johannes Stille:	for Net-2Debugged by 
 *					<johannes@titan.os.open.de>
 *	 	Linus Torvalds:		Misc Changes
 *	 	Alan Cox:		add the new mtu/window stuff
 *	 	Miquel van Smoorenburg:	rt_add and rt_del
 *	 {1.79}	Bernd Eckenfels:	route_info
 *	 {1.80}	Bernd Eckenfels:	reject, metric, irtt, 1.2.x support.
 *	 {1.81}	Bernd Eckenfels:	reject routes need a dummy device
 *960127 {1.82} Bernd Eckenfels:	'mod' and 'dyn' 'reinstate' added
 *960129 {1.83}	Bernd Eckenfels:	resolve and getsock now in lib/, 
 *					REJECT displays '-' as gatway.
 *960202 {1.84} Bernd Eckenfels:	net-features support added
 *960203 {1.85} Bernd Eckenfels:	"#ifdef' in '#if' for net-features
 *					-A  (aftrans) support, get_longopts
 *960206 {1.86} Bernd Eckenfels:	route_init();
 *960218 {1.87} Bernd Eckenfels:	netinet/in.h added
 *960221 {1.88}	Bernd Eckenfels:	aftrans_dfl support
 *960222 {1.90} Bernd Eckenfels:	moved all AF specific code to lib/.
 *960413 {1.91} Bernd Eckenfels:	new RTACTION support+FLAG_CACHE/FIB
 *960426 {1.92} Bernd Eckenfels:	FLAG_SYM/-N support
 *
 */
#include <sys/types.h>
#include <sys/ioctl.h>
#if defined(__GLIBC__)
#define _SOCKETBITS_H
#endif /* __GLIBC__ */
#include <sys/socket.h>
#if !defined(__GLIBC__)
#include <net/if.h>
#include <netinet/in.h>
#else /* __GLIBC__ */
#include <linux/in.h>
#define _NETINET_IN_H
#endif /* __GLIBC__ */
#include <netdb.h>
#include <netinet/in.h>
#include <arpa/nameser.h>
#include <resolv.h>
#include <linux/param.h>
#include <stdio.h>
#include <errno.h>
#include <fcntl.h>
#include <stdlib.h>
#include <string.h>
#include <getopt.h>
#include <unistd.h>
#include <ctype.h>
#include <linux/route.h>
#include "net-support.h"
#include "config.h"
#include "net-locale.h"
#include "pathnames.h"
#include "version.h"

#define DFLT_AF "inet"

#define FEATURE_ROUTE
#include "lib/net-features.h" /* needs some of the system includes above! */

char	*Release = RELEASE,
	*Version = "route 1.92 (1996-04-26)";

int             opt_n  = 0;	/* numerical output flag	*/
int             opt_v  = 0;	/* debugging output flag	*/
int		opt_e  = 1;	/* 1,2,3=type of routetable	*/
int		opt_fc = 0;	/* routing cache/FIB */
int		opt_h  = 0;	/* help selected		*/
struct aftype   *ap;		/* current address family	*/

static void usage(void)
{
	fprintf(stderr, NLS_CATGETS(catfd, routeSet, route_usage1,
				"Usage: route [-nNvee] [-FC] [Address_families]  List kernel routing tables\n"));
	fprintf(stderr, NLS_CATGETS(catfd, routeSet, route_usage2,
				"       route {-V|--version}                  Display command version and exit.\n"));
	fprintf(stderr, NLS_CATGETS(catfd, routeSet, route_usage3,
				"       route {-h|--help} [Address_family]    Usage Syntax for specified AF.\n"));
	fprintf(stderr, NLS_CATGETS(catfd, routeSet, route_usage4,
				"       route [-v] [-FC] {add|del|flush} ...  Modify routing table for AF.\n\n"));
	fprintf(stderr, NLS_CATGETS(catfd, routeSet, route_usage5,
				"  Address_families: inet,ddp,ipx,netrom,ax25\n"));
	fprintf(stderr, NLS_CATGETS(catfd, routeSet, route_usage6,
				"        specify AF: -A af1,af2..  or  --af1 --af2  or  af_route\n"));
	NLS_CATCLOSE(catfd)
	exit(E_USAGE);
}


static void version(void)
{
	fprintf(stderr,"%s\n%s\n%s\n",Release,Version,Features);
	NLS_CATCLOSE(catfd)
	exit(E_VERSION);
}


int main(int argc, char **argv)
{
	int i,lop,what=0;
	struct option longopts[]=
	{
		AFTRANS_OPTS,
		{"extend",	0,	0,	'e'},
		{"verbose",	0,	0,	'v'},
		{"version",	0,	0,	'V'},
		{"numeric",	0,	0,	'n'},
		{"symbolic",	0,	0,	'N'},
		{"protocol",	1,	0,	'A'},
		{"cache",	0,	0,	'C'},
		{"fib",		0,	0,	'F'},
		{"help",	0,	0,	'h'},
		{NULL,		0,	0,	0}
	};	
	char **tmp;
	char *progname;
	int options;

#if NLS
	setlocale (LC_MESSAGES, "");
	catfd = catopen ("nettools", MCLoadBySet);
#endif

	getroute_init(); 			/* Set up AF routing support */
	setroute_init();
	afname[0] = '\0';
	progname=argv[0];
	
	/* getopts and -net wont work :-/ */
	for(tmp=argv;*tmp;tmp++) {
		if (!strcmp(*tmp, "-net"))
			strcpy(*tmp, "#net");
		else if (!strcmp(*tmp, "-host"))
			strcpy(*tmp, "#host");
	}
	
	/* Fetch the command-line arguments. */
	while ((i = getopt_long(argc, argv, "A:eCFhnNVv?", longopts, &lop)) != EOF) switch(i) {
		case -1:
			break;
		case 'n':
			opt_n |= FLAG_NUM;
			break;
		case 'N':
			opt_n |= FLAG_SYM;
			break;
		case 'v':
			opt_v |= FLAG_VERBOSE;
			break;
		case 'e':
			opt_e++;
			break;
		case 1:
			if (lop < 0 || lop >= AFTRANS_CNT) {
				EINTERN("route.c","longopts 1 range");
				break;
			}
			if ((i = aftrans_opt(longopts[lop].name))) {
				NLS_CATCLOSE(catfd)
				exit(i);
			}
			break;
		case 'C':
			opt_fc |= FLAG_CACHE;
			break;
		case 'F':
			opt_fc |= FLAG_FIB;
			break;
		case 'A':
			if ((i = aftrans_opt(optarg))) {
				NLS_CATCLOSE(catfd)
				exit(i);
			}
			break;
		case 'V':
			version();
		case 'h':
		case '?':
			opt_h++;
			break;
		default:
			usage();
	}
	
	argv += optind;
	argc -= optind;
	
	if (opt_h) {
		if (!afname[0])
			usage();
		else
			what = RTACTION_HELP;
	} else {	
		if (!afname[0])
			/* this will initialise afname[] */
			aftrans_def("route", progname, DFLT_AF);

		/* Do we have to show the contents of the routing table? */
		if (*argv == NULL) {
			what = RTACTION_SHOW;
		} else {
			if (!strcmp(*argv, "add"))
				what = RTACTION_ADD;
			else if (!strcmp(*argv, "del") || !strcmp(*argv, "delete"))
				what = RTACTION_DEL;
			else if (!strcmp(*argv,"flush"))
				what = RTACTION_FLUSH;
			else 
				usage();
		}
	}

	options = (opt_e & FLAG_EXT) | opt_n | opt_fc | opt_v;
	if (!opt_fc)
		options |= FLAG_FIB;

	if (what == RTACTION_SHOW)
		i = route_info(afname,options);
	else
		i = route_edit(what,afname,options,++argv);

	if (i == E_OPTERR)
		usage();

	NLS_CATCLOSE(catfd)
	return (i);
}
